EToCalculator.ts

```diff
@@ -0,0 +1,62 @@
+import { EToData } from "./types";
+
+// The implementation of this algorithm was guided by a step-by-step breakdown (http://edis.ifas.ufl.edu/pdffiles/ae/ae45900.pdf)
+/**
+ * Calculates the reference evapotranspiration using the Penman-Monteith (FAO-56) method (http://www.fao.org/3/X0490E/x0490e07.htm).
+ *
+ * @param etoData The data to calculate the ETo with.
+ * @param elevation The elevation above sea level of the watering site (in meters).
+ * @return The reference evapotranspiration (in millimeters per day).
+ */
+export function calculateETo( etoData: EToData, elevation: number ): number {
+	const avgTemp = ( etoData.maxTemp + etoData.minTemp ) / 2;
+
+	// Adjust the wind speed to a 2m height.
+	const windSpeed = etoData.windSpeed * 4.87 / Math.log( 67.8 * etoData.windSpeedMeasurementHeight - 5.42 );
+
+	const saturationVaporPressureCurveSlope = 4098 * 0.6108 * Math.exp( 17.27 * avgTemp / ( avgTemp + 237.3 ) ) / Math.pow( avgTemp + 237.3, 2 );
+
+	const pressure = 101.3 * Math.pow( ( 293 - 0.0065 * elevation ) / 293, 2.36 );
+
+	const psychrometricConstant = 0.000665 * pressure;
+
+	const deltaTerm = saturationVaporPressureCurveSlope / ( saturationVaporPressureCurveSlope + psychrometricConstant * ( 1 + 0.34 * windSpeed ) );
+
+	const psiTerm = psychrometricConstant / ( saturationVaporPressureCurveSlope + psychrometricConstant * ( 1 + 0.34 * windSpeed ) );
+
+	const tempTerm = ( 900 / ( avgTemp + 273 ) ) * windSpeed;
+
+	const minSaturationVaporPressure = 0.6108 * Math.exp( 17.27 * etoData.minTemp / ( etoData.minTemp + 237.3 ) );
+
+	const maxSaturationVaporPressure = 0.6108 * Math.exp( 17.27 * etoData.maxTemp / ( etoData.maxTemp + 237.3 ) );
+
+	const avgSaturationVaporPressure = ( minSaturationVaporPressure + maxSaturationVaporPressure ) / 2;
+
+	const actualVaporPressure = ( minSaturationVaporPressure * etoData.maxHumidity / 100 + maxSaturationVaporPressure * etoData.minHumidity / 100 ) / 2;
```

Comment:
Very minor point, DarkSky Weather Provider is approximating daily min/max humidity from hourly values. Not sure if that is better than using the PM Step 11b equation with DarkSky's mean daily humidity.

Reply:
The min/max humidity values we're calculating should still be pretty close to the actual values. I think we should leave it as-is for now, but I'll look into changing it in the near future when adding support to other WeatherProviders if they won't be as accurate (e.g. OWM offers data in 3 hour windows instead of 1 hour windows, so this method of determining min/max humidity won't be as accurate).

Reply:
Resolved in favor of #56 .